File: plugins/css-duplication/lsp.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
)

// A language server, not a hook: diagnostics arrive in context on their own
// after an edit instead of a message shouted at the end of a tool call. The
// detector is shared with nothing else to keep the two in agreement -- css.go
// is the single implementation.

type rpcMessage struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      json.RawMessage `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  any             `json:"result,omitempty"`
	Error   *rpcError       `json:"error,omitempty"`
}

type rpcError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

type position struct {
	Line      int `json:"line"`      // 0-based
	Character int `json:"character"` // 0-based, UTF-16 units
}

type rng struct {
	Start position `json:"start"`
	End   position `json:"end"`
}

type location struct {
	URI   string `json:"uri"`
	Range rng    `json:"range"`
}

type relatedInfo struct {
	Location location `json:"location"`
	Message  string   `json:"message"`
}

type diagnostic struct {
	Range              rng           `json:"range"`
	Severity           int           `json:"severity"` // 2 = Warning
	Source             string        `json:"source"`
	Message            string        `json:"message"`
	RelatedInformation []relatedInfo `json:"relatedInformation,omitempty"`
}

type textDocumentItem struct {
	URI  string `json:"uri"`
	Text string `json:"text"`
}

// Server holds the open documents. A language server is long-lived and is
// spoken to concurrently in principle, so the map is guarded even though the
// client drives it from one connection.
type Server struct {
	out io.Writer

	mu   sync.Mutex
	docs map[string]string // uri -> text
}

func NewServer(out io.Writer) *Server {
	return &Server{out: out, docs: map[string]string{}}
}

const (
	severityWarning  = 2
	diagnosticSource = "css-duplication"
)

// stylesheetExts is deliberately narrow, and matches what .lsp.json registers.
// A preprocessor's nesting changes what a duplicate body MEANS (`&:hover`
// under two parents is not a repeated rule), so this parser only claims plain
// CSS.
var stylesheetExts = map[string]bool{".css": true}

// Serve runs the stdio JSON-RPC loop until the stream ends or `exit` arrives.
func (s *Server) Serve(in io.Reader) error {
	r := bufio.NewReader(in)
	for {
		body, err := readMessage(r)
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}
		var msg rpcMessage
		if err := json.Unmarshal(body, &msg); err != nil {
			continue // a malformed frame is not worth killing the server over
		}
		stop, result, rerr := s.dispatch(msg)
		if len(msg.ID) > 0 { // a request: it must be answered, even with null
			s.respond(msg.ID, result, rerr)
		}
		if stop {
			return nil
		}
	}
}

func (s *Server) dispatch(msg rpcMessage) (stop bool, result any, rerr *rpcError) {
	switch msg.Method {
	case "initialize":
		return false, initializeResult(), nil

	case "initialized", "workspace/didChangeConfiguration", "$/setTrace":
		return false, nil, nil

	case "textDocument/didOpen":
		var p struct {
			TextDocument textDocumentItem `json:"textDocument"`
		}
		if err := json.Unmarshal(msg.Params, &p); err == nil {
			s.setDoc(p.TextDocument.URI, p.TextDocument.Text)
			s.publish(p.TextDocument.URI)
		}
		return false, nil, nil

	case "textDocument/didChange":
		// Full sync only (that is what the capabilities advertise), so the last
		// content change is the whole document.
		var p struct {
			TextDocument   textDocumentItem `json:"textDocument"`
			ContentChanges []struct {
				Text string `json:"text"`
			} `json:"contentChanges"`
		}
		if err := json.Unmarshal(msg.Params, &p); err == nil && len(p.ContentChanges) > 0 {
			s.setDoc(p.TextDocument.URI, p.ContentChanges[len(p.ContentChanges)-1].Text)
			s.publish(p.TextDocument.URI)
		}
		return false, nil, nil

	case "textDocument/didSave":
		var p struct {
			TextDocument textDocumentItem `json:"textDocument"`
			Text         string           `json:"text"`
		}
		if err := json.Unmarshal(msg.Params, &p); err == nil {
			if p.Text != "" {
				s.setDoc(p.TextDocument.URI, p.Text)
			}
			s.publish(p.TextDocument.URI)
		}
		return false, nil, nil

	case "textDocument/didClose":
		var p struct {
			TextDocument textDocumentItem `json:"textDocument"`
		}
		if err := json.Unmarshal(msg.Params, &p); err == nil {
			s.mu.Lock()
			delete(s.docs, p.TextDocument.URI)
			s.mu.Unlock()
			// Clear what we published, so a closed file leaves no ghosts.
			s.notify("textDocument/publishDiagnostics", map[string]any{
				"uri": p.TextDocument.URI, "diagnostics": []diagnostic{},
			})
		}
		return false, nil, nil

	// Pull diagnostics: a client that prefers asking gets the same answer the
	// push path would have sent.
	case "textDocument/diagnostic":
		var p struct {
			TextDocument textDocumentItem `json:"textDocument"`
		}
		if err := json.Unmarshal(msg.Params, &p); err != nil {
			return false, nil, &rpcError{Code: -32602, Message: "invalid params"}
		}
		return false, map[string]any{
			"kind":  "full",
			"items": s.diagnose(p.TextDocument.URI),
		}, nil

	case "shutdown":
		return false, nil, nil

	case "exit":
		return true, nil, nil
	}

	if len(msg.ID) > 0 {
		return false, nil, &rpcError{Code: -32601, Message: "method not found: " + msg.Method}
	}
	return false, nil, nil
}

func initializeResult() map[string]any {
	return map[string]any{
		"capabilities": map[string]any{
			// 1 = full document sync. The detector needs the whole stylesheet
			// anyway (a duplicate is a relationship between distant rules), so
			// incremental sync would buy nothing but bookkeeping.
			"textDocumentSync": map[string]any{
				"openClose": true,
				"change":    1,
				"save":      map[string]any{"includeText": true},
			},
			"diagnosticProvider": map[string]any{
				"identifier":            diagnosticSource,
				"interFileDependencies": false,
				"workspaceDiagnostics":  false,
			},
		},
		"serverInfo": map[string]any{"name": "css-duplication", "version": "1"},
	}
}

func (s *Server) setDoc(uri, text string) {
	s.mu.Lock()
	s.docs[uri] = text
	s.mu.Unlock()
}

func (s *Server) publish(uri string) {
	s.notify("textDocument/publishDiagnostics", map[string]any{
		"uri":         uri,
		"diagnostics": s.diagnose(uri),
	})
}

// diagnose is the whole point: one diagnostic per copy, each naming the other
// selectors carrying the identical block, with the fix in the message.
func (s *Server) diagnose(uri string) []diagnostic {
	s.mu.Lock()
	text, ok := s.docs[uri]
	s.mu.Unlock()
	if !ok || !isStylesheet(uri) {
		return []diagnostic{}
	}

	lines := strings.Split(text, "\n")
	out := []diagnostic{}
	for _, g := range FindDuplicates(ParseRules(text)) {
		body := "{ " + strings.Join(g.Decls, "; ") + " }"
		for i, r := range g.Rules {
			var related []relatedInfo
			for _, other := range g.Rules {
				if other.Line == r.Line && other.Selector == r.Selector {
					continue
				}
				related = append(related, relatedInfo{
					Location: location{URI: uri, Range: selectorRange(lines, other)},
					Message:  "same block here: " + other.Selector,
				})
			}
			out = append(out, diagnostic{
				Range:    selectorRange(lines, r),
				Severity: severityWarning,
				Source:   diagnosticSource,
				// Terse on purpose, and said in full exactly ONCE per finding.
				// The client concatenates diagnostics into one block, keeps ten
				// per file and truncates at 4000 characters, so restating the
				// body and the sibling list on every copy would spend a
				// finding's whole budget on repeating itself. The first copy
				// carries the detail; the rest point at it, which is also how a
				// human reads it -- one explanation, several markers.
				Message:            groupMessage(g, i, body),
				RelatedInformation: related,
			})
		}
	}
	return out
}

// maxNamedSiblings bounds the "also on" list. A table rule can share its body
// with a dozen selectors, and the tail of that list teaches nothing the count
// does not -- while it does crowd out the next finding.
const maxNamedSiblings = 3

// groupMessage writes the finding once, on its first copy, and makes every
// later copy a pointer back to it.
func groupMessage(g Group, i int, body string) string {
	if i > 0 {
		return fmt.Sprintf("same block as line %d -- %d copies; hoist to one rule (/docs:css-cascade)",
			g.Rules[0].Line, len(g.Rules))
	}
	others := make([]string, 0, len(g.Rules)-1)
	for _, other := range g.Rules[1:] {
		others = append(others, other.Selector)
	}
	named := others
	if len(named) > maxNamedSiblings {
		named = append(append([]string(nil), others[:maxNamedSiblings]...),
			fmt.Sprintf("+%d more", len(others)-maxNamedSiblings))
	}
	return fmt.Sprintf("%s also on %s -- %d copies; hoist to one rule (/docs:css-cascade)",
		body, strings.Join(named, ", "), len(g.Rules))
}

// selectorRange points at the selector text itself when it can be found on its
// line, and at the whole line otherwise -- a diagnostic with a bogus range is
// worse than a coarse one.
func selectorRange(lines []string, r Rule) rng {
	idx := r.Line - 1
	if idx < 0 || idx >= len(lines) {
		return rng{}
	}
	line := lines[idx]
	// A multi-selector rule is stored whitespace-collapsed, so match its first
	// token rather than the reassembled string.
	needle := r.Selector
	if i := strings.IndexAny(needle, ",\n"); i > 0 {
		needle = needle[:i]
	}
	start := strings.Index(line, needle)
	if start < 0 {
		return rng{Start: position{Line: idx}, End: position{Line: idx, Character: len(line)}}
	}
	return rng{
		Start: position{Line: idx, Character: start},
		End:   position{Line: idx, Character: start + len(needle)},
	}
}

func isStylesheet(uri string) bool {
	path := uri
	if u, err := url.Parse(uri); err == nil && u.Scheme == "file" {
		path = u.Path
	}
	return stylesheetExts[strings.ToLower(filepath.Ext(path))]
}

func (s *Server) notify(method string, params any) {
	raw, err := json.Marshal(params)
	if err != nil {
		return
	}
	s.send(rpcMessage{JSONRPC: "2.0", Method: method, Params: raw})
}

// respond answers a request. `result` is written even when it is nil, as an
// explicit JSON null: a success response carrying NEITHER result nor error is
// malformed, and a real client rejects it -- vscode-jsonrpc fails the request
// with "The received response has neither a result nor an error property",
// which is how `shutdown` was breaking after a clean, fully working session.
// Struct tags cannot express this (`omitempty` drops the null, dropping it is
// the bug), so the response is assembled by hand.
func (s *Server) respond(id json.RawMessage, result any, rerr *rpcError) {
	out := map[string]any{"jsonrpc": "2.0", "id": id}
	if rerr != nil {
		out["error"] = rerr
	} else {
		out["result"] = result // nil marshals to null, which is what LSP wants
	}
	body, err := json.Marshal(out)
	if err != nil {
		return
	}
	fmt.Fprintf(s.out, "Content-Length: %d\r\n\r\n%s", len(body), body)
}

func (s *Server) send(msg rpcMessage) {
	msg.JSONRPC = "2.0"
	body, err := json.Marshal(msg)
	if err != nil {
		return
	}
	fmt.Fprintf(s.out, "Content-Length: %d\r\n\r\n%s", len(body), body)
}

// readMessage reads one LSP frame: headers, blank line, then Content-Length
// bytes of body.
func readMessage(r *bufio.Reader) ([]byte, error) {
	length := -1
	for {
		line, err := r.ReadString('\n')
		if err != nil {
			return nil, err
		}
		line = strings.TrimRight(line, "\r\n")
		if line == "" {
			break
		}
		name, value, ok := strings.Cut(line, ":")
		if !ok {
			continue
		}
		if strings.EqualFold(strings.TrimSpace(name), "content-length") {
			n, err := strconv.Atoi(strings.TrimSpace(value))
			if err != nil {
				return nil, fmt.Errorf("bad Content-Length: %q", value)
			}
			length = n
		}
	}
	if length < 0 {
		return nil, fmt.Errorf("frame without Content-Length")
	}
	body := make([]byte, length)
	if _, err := io.ReadFull(r, body); err != nil {
		return nil, err
	}
	return body, nil
}

File: src/lsp.ts
// A stdio language server whose only job is to say, while a file is open, what
// common-checks would say about it in CI.

import { findings, fileKind, type Finding } from "./checks.ts";

export const SOURCE = "common-checks";

// The client injects only the first handful of diagnostics per file, so a file
// with more findings than that says so on the last one it does send. Silently
// dropping the tail would read as "that is all of it".
export const MAX_PER_FILE = Number(process.env.COMMON_CHECKS_LSP_MAX_PER_FILE) || 10;

export interface Diagnostic {
  range: { start: { line: number; character: number }; end: { line: number; character: number } };
  severity: number;
  source: string;
  code: string;
  message: string;
}

/**
 * Every diagnostic is severity 1 (Error): a finding here is a finding that
 * fails the merge gate. ste-lint's heuristic buckets never reach this point --
 * they do not fail CI, and spending the client's budget on them would push out
 * findings that do.
 */
export function toDiagnostic(finding: Finding, lines: string[]): Diagnostic {
  const start = Math.max(0, finding.startLine - 1);
  const end = Math.max(start, finding.endLine - 1);
  const width = lines[end]?.length ?? 0;
  return {
    range: { start: { line: start, character: 0 }, end: { line: end, character: width } },
    severity: 1,
    source: SOURCE,
    code: finding.check,
    message: finding.message,
  };
}

export function diagnosticsFor(relativePath: string, content: string): Diagnostic[] {
  const all = findings(relativePath, content);
  const lines = content.split(/\r?\n/);
  const shown = all.slice(0, MAX_PER_FILE).map((finding) => toDiagnostic(finding, lines));
  const hidden = all.length - shown.length;
  if (hidden > 0 && shown.length > 0) {
    shown[shown.length - 1].message += ` (+${hidden} more common-checks findings in this file)`;
  }
  return shown;
}

/** Decodes a `file://` URI to a filesystem path. */
export function pathOf(uri: string): string {
  if (!uri.startsWith("file://")) return uri;
  try {
    return decodeURIComponent(uri.slice("file://".length));
  } catch {
    return uri.slice("file://".length);
  }
}

/**
 * The path the checks reason about, which is always repository-relative. When
 * the client named a root, the path is taken against it. Otherwise the
 * `.github/workflows/` segment locates it, and everything else falls back to a
 * basename -- enough for an `action.yml` or a `.md`, which are recognised by
 * their name alone.
 */
export function relativize(uri: string, root: string | undefined): string {
  const path = pathOf(uri).replace(/\\/g, "/");
  if (root) {
    const base = pathOf(root).replace(/\\/g, "/").replace(/\/+$/, "");
    if (path.startsWith(`${base}/`)) return path.slice(base.length + 1);
  }
  const workflows = path.lastIndexOf("/.github/workflows/");
  if (workflows !== -1) return path.slice(workflows + 1);
  const slash = path.lastIndexOf("/");
  return slash === -1 ? path : path.slice(slash + 1);
}

interface Message {
  id?: number | string | null;
  method?: string;
  params?: Record<string, unknown>;
}

type Write = (chunk: string) => void;

export class Server {
  #root: string | undefined;
  #open = new Map<string, string>();

  constructor(private readonly write: Write) {}

  /** Frames one JSON-RPC message the way the base protocol requires. */
  #send(payload: Record<string, unknown>): void {
    const body = JSON.stringify(payload);
    this.write(`Content-Length: ${Buffer.byteLength(body, "utf8")}\r\n\r\n${body}`);
  }

  // Assembled by hand so a null result marshals as an explicit `null`. A reply
  // carrying neither `result` nor `error` is malformed JSON-RPC, and the
  // client's reader rejects it -- which is how a clean session still fails at
  // shutdown.
  #respond(id: number | string, result: unknown): void {
    this.#send({ jsonrpc: "2.0", id, result: result === undefined ? null : result });
  }

  #publish(uri: string): void {
    const content = this.#open.get(uri);
    if (content === undefined) return;
    const relative = relativize(uri, this.#root);
    const diagnostics = fileKind(relative) === "other" ? [] : diagnosticsFor(relative, content);
    this.#send({
      jsonrpc: "2.0",
      method: "textDocument/publishDiagnostics",
      params: { uri, diagnostics },
    });
  }

  handle(message: Message): void {
    const { id, method, params } = message;
    switch (method) {
      case "initialize": {
        const folders = params?.workspaceFolders as { uri?: string }[] | undefined;
        this.#root = (params?.rootUri as string | undefined) ?? folders?.[0]?.uri;
        this.#respond(id as number, {
          // Full sync: a finding is a property of the whole document, and an
          // incremental edit would need the server to rebuild it anyway.
          capabilities: { textDocumentSync: 1, diagnosticProvider: { interFileDependencies: false, workspaceDiagnostics: false } },
          serverInfo: { name: SOURCE, version: "1" },
        });
        return;
      }
      case "initialized":
        return;
      case "shutdown":
        this.#respond(id as number, null);
        return;
      case "exit":
        return;
      case "textDocument/didOpen": {
        const doc = params?.textDocument as { uri?: string; text?: string } | undefined;
        if (!doc?.uri) return;
        this.#open.set(doc.uri, doc.text ?? "");
        this.#publish(doc.uri);
        return;
      }
      case "textDocument/didChange": {
        const doc = params?.textDocument as { uri?: string } | undefined;
        const changes = params?.contentChanges as { text?: string }[] | undefined;
        const text = changes?.[changes.length - 1]?.text;
        if (!doc?.uri || text === undefined) return;
        this.#open.set(doc.uri, text);
        this.#publish(doc.uri);
        return;
      }
      case "textDocument/didSave": {
        const doc = params?.textDocument as { uri?: string } | undefined;
        if (doc?.uri) this.#publish(doc.uri);
        return;
      }
      case "textDocument/didClose": {
        const doc = params?.textDocument as { uri?: string } | undefined;
        if (!doc?.uri) return;
        this.#open.delete(doc.uri);
        // Clearing is what removes the findings from a file nobody is editing.
        this.#send({ jsonrpc: "2.0", method: "textDocument/publishDiagnostics", params: { uri: doc.uri, diagnostics: [] } });
        return;
      }
      // Claude Code never calls this -- it registers a publishDiagnostics
      // handler and nothing else. It is here so the server is correct for an
      // editor that pulls.
      case "textDocument/diagnostic": {
        const doc = params?.textDocument as { uri?: string } | undefined;
        const content = doc?.uri ? this.#open.get(doc.uri) : undefined;
        const relative = doc?.uri ? relativize(doc.uri, this.#root) : "";
        const items =
          content === undefined || fileKind(relative) === "other" ? [] : diagnosticsFor(relative, content);
        this.#respond(id as number, { kind: "full", items });
        return;
      }
      default:
        // A request must always be answered; a notification must not be.
        if (id !== undefined && id !== null) {
          this.#send({ jsonrpc: "2.0", id, error: { code: -32601, message: `unknown method: ${method}` } });
        }
    }
  }
}

/** Splits the base protocol's header-framed stream into messages. */
export class Framing {
  #buffer = Buffer.alloc(0);

  push(chunk: Buffer): Message[] {
    this.#buffer = Buffer.concat([this.#buffer, chunk]);
    const out: Message[] = [];
    for (;;) {
      const split = this.#buffer.indexOf("\r\n\r\n");
      if (split === -1) return out;
      const header = this.#buffer.subarray(0, split).toString("ascii");
      const length = /content-length:\s*(\d+)/i.exec(header);
      if (!length) {
        // A frame with no length cannot be skipped safely, so drop the header
        // and resynchronise rather than blocking on a stream that will never
        // produce one.
        this.#buffer = this.#buffer.subarray(split + 4);
        continue;
      }
      const size = Number(length[1]);
      const start = split + 4;
      if (this.#buffer.length < start + size) return out;
      const body = this.#buffer.subarray(start, start + size).toString("utf8");
      this.#buffer = this.#buffer.subarray(start + size);
      try {
        out.push(JSON.parse(body) as Message);
      } catch {
        // A frame that is not JSON is one message lost, never the session.
      }
    }
  }
}

export function serve(input: NodeJS.ReadableStream, write: Write): void {
  const framing = new Framing();
  const server = new Server(write);
  input.on("data", (chunk: Buffer) => {
    for (const message of framing.push(chunk)) {
      try {
        server.handle(message);
      } catch (error) {
        // One bad document must not take the session down with it: the editor
        // would lose every later finding and say nothing about why.
        process.stderr.write(`common-checks: ${error instanceof Error ? error.stack : String(error)}\n`);
      }
    }
  });
}
